src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,400 @@
+                    }
+                }
+
+                // This may well be the first time that we've added something to this store.
+                Directory.CreateDirectory(_storePath);
+
+                string thumbprint = copy.Thumbprint;
+                bool findOpenSlot;
+
+                // The odds are low that we'd have a thumbprint colission, but check anyways.
+                string existingFilename = FindExistingFilename(copy, _storePath, out findOpenSlot);
+
+                if (existingFilename != null)
+                {
+                    bool dataExistsAlready = false;
+
+                    // If the file on disk is just a public key, but we're trying to add a private key,
+                    // we'll want to overwrite it.
+                    if (copy.HasPrivateKey)
+                    {
+                        try
+                        {
+                            using (X509Certificate2 fromFile = new X509Certificate2(existingFilename))
+                            {
+                                if (fromFile.HasPrivateKey)
+                                {
+                                    // We have a private key, the file has a private key, we're done here.
+                                    dataExistsAlready = true;
+                                }
+                            }
+                        }
+                        catch (CryptographicException)
+                        {
+                            // We can't read this file anymore, but a moment ago it was this certificate,
+                            // so go ahead and overwrite it.
+                        }
+                    }
+                    else
+                    {
+                        // If we're just a public key then the file has at least as much data as we do.
+                        dataExistsAlready = true;
+                    }
+
+                    if (dataExistsAlready)
+                    {
+                        // The file was added but our collection hasn't resynced yet.
+                        // Set _certificates to null to force a resync.
+                        _certificates = null;
+                        return;
+                    }
+                }
+
+                string destinationFilename;
+
+                if (existingFilename != null)
+                {
+                    destinationFilename = existingFilename;
+                }
+                else if (findOpenSlot)
+                {
+                    destinationFilename = FindOpenSlot(thumbprint);
+                }
+                else
+                {
+                    destinationFilename = Path.Combine(_storePath, thumbprint + PfxExtension);
+                }
+
+                // With no explicit permissions dictated these are being created with Unix permissions of
+                // -rwx------
+                // which leaves the private keys scoped to the user, and makes us happy.
```

Comment:
If this is important, should we assert that it remains true?  These were the permissions I used at https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/UnixFileStream.cs#L88, but if that's important for this code, we'll want to make sure we don't inadvertently breaking it by changing those permissions.  Plus, it's possible for the system to use a umask to further constrain the permissions when new files are created; not sure if that negatively affects this.